crates/cxx-qt/src/lib.rs

```diff
@@ -50,3 +50,118 @@ pub trait CxxQtType {
 pub trait Locking {
     // empty
 }
+
+/// This trait can be implemented inside the [`#[cxx_qt::bridge]`](bridge) on any [CxxQtType] to define a
+/// custom constructor in C++ for the QObject.
+///
+/// The `Arguments` must be a tuple of CXX types that will be the arguments to the constructor in C++.
+///
+/// Note that implementing this trait will stop Cxx-Qt from generating a default constructor.
+///
+/// # Minimal Example
+/// ```no_run
+/// # #[macro_use] extern crate cxx;
+/// #[cxx_qt::bridge]
+/// mod ffi {
+///     // Struct without `Default` implementation
+///     #[cxx_qt::qobject]
+///     pub struct MyStruct {
+///         integer: i32,
+///         string: String
+///     }
+///
+///     impl cxx_qt::Constructor<(i32, String)> for qobject::MyStruct {
+///         type BaseArguments = (); // Will be passed to the base class constructor
+///         type InitializeArguments = (); // Will be passed to the "initialize" function
+///         type NewArguments = (i32, String); // Will be passed to the "new" function
+///
+///         fn route_arguments(args: (i32, String)) -> (
+///             Self::NewArguments,
+///             Self::BaseArguments,
+///             Self::InitializeArguments
+///         ) {
+///             (args, (), ())
+///         }
+///
+///         fn new((integer, string): (i32, String)) -> MyStruct {
+///             MyStruct {
+///                 integer,
+///                 string
+///             }
+///         }
+///     }
+/// }
+///
+/// # fn main() {}
+/// ```
+///
+/// # Pseudo Code for generated C++ Constructor
+/// You can imagine this trait as creating a constructor roughly like this:
+/// ``` cpp
+/// class MyCxxQtType : public QObject {
+///     public:
+///         MyCxxQtType(Arguments... args)
+///             : QObject(route_arguments(args).BaseArguments)
+///             , m_rust(Constructor::new(route_arguments(args).NewArguments))
+///         {
+///             Constructor::initialize(*this, route_arguments(args).InitializeArguments);
+///         }
+/// }
+/// ```
+/// Note that in reality, `route_arguments` will only be called once and all arguments
+/// will be moved, never copied.
+pub trait Constructor<Arguments>: CxxQtType {
+    /// The arguments that are passed to the [`new()`](Self::new) function to construct the inner Rust struct.
+    /// This must be a tuple of Cxx compatible types.
+    ///
+    /// This way QObjects can be constructed that need additional arguments for constructing the
+    /// inner Rust type.
+    // TODO: Enable Defaults
+    // # Default
+    // If nothing is specified, CxxQt will default this to `Arguments`.
+    type NewArguments;
+    /// The arguments that should be passed to the constructor of the base class.
+    /// This must be a tuple of Cxx compatible types.
+    // TODO: Enable Defaults
```

Comment:
Is this going to happen in this pull request or future ones?

Reply:
In future ones, I just wanted to document here what the plan was.
Note that it's using double `/`, so it's a normal comment, not part of the documentation (yet).